File: e2e-test.mjs
// E2E test: agent draw -> render plan like the browser does -> upload ->
// analysis stages -> verify DB row has real (non-blank) image.
// Usage: node e2e-test.mjs "a small sailboat on calm water at sunset"
import { createCanvas } from 'node:canvas';

const BASE = process.env.BASE || 'http://127.0.0.1:8910';
const API_KEY = process.env.DRAW_API_KEY;

const subject = process.argv[2] || 'a small sailboat on calm water at sunset';
const headers = { 'content-type': 'application/json', 'x-api-key': API_KEY };

async function j(path, opts) {
  const r = await fetch(BASE + path, { ...opts, headers: { ...headers, ...(opts?.headers || {}) } });
  const body = await r.json().catch(() => ({}));
  return { status: r.status, ok: r.ok, body };
}

// --- 1. create a draw run
const create = await j('/api/draw', {
  method: 'POST',
  body: JSON.stringify({ prompt: subject, stages: ['describe', 'critique'] }),
});
console.log('1. create:', create.status, JSON.stringify(create.body));
if (!create.ok) process.exit(1);
const id = create.body.run_id;

// --- 2. poll until rendering (plan ready)
let plan = null;
const t0 = Date.now();
while (Date.now() - t0 < 360000) {
  const r = await j(`/api/runs/${id}`);
  if (r.body.status === 'error') { console.log('PLAN ERROR:', r.body.error); process.exit(1); }
  if (r.body.status === 'rendering') {
    const ds = (r.body.stages || []).find((s) => s.key === 'draw');
    if (ds && ds.plan) plan = ds.plan;
    break;
  }
  await new Promise((res) => setTimeout(res, 3000));
}
if (!plan) { console.log('no plan after 6 min'); process.exit(1); }
console.log(`2. plan ready in ${Math.round((Date.now() - t0) / 1000)}s:`, plan.count, 'strokes, subject:', plan.subject);

// --- 3. render the plan (same code path the browser uses)
const canvas = createCanvas(960, 600);
const ctx = canvas.getContext('2d');
ctx.fillStyle = '#ffffff';
ctx.fillRect(0, 0, 960, 600);
ctx.lineCap = 'round';
ctx.lineJoin = 'round';
for (const s of plan.strokes) {
  ctx.strokeStyle = s.color;
  ctx.fillStyle = s.color;
  ctx.lineWidth = s.width || 3;
  if (s.type === 'line') {
    ctx.beginPath();
    ctx.moveTo(s.points[0][0], s.points[0][1]);
    for (const p of s.points) ctx.lineTo(p[0], p[1]);
    ctx.stroke();
  } else if (s.type === 'circle') {
    ctx.beginPath();
    ctx.arc(s.cx, s.cy, s.r, 0, Math.PI * 2);
    if (s.fill) ctx.fill(); else ctx.stroke();
  } else if (s.type === 'rect') {
    ctx.beginPath();
    ctx.rect(s.x, s.y, s.w, s.h);
    if (s.fill) ctx.fill(); else ctx.stroke();
  } else if (s.type === 'text') {
    ctx.font = `${s.size || 24}px sans-serif`;
    canvas.getContext('2d').fillText(s.text, s.x, s.y);
  }
}
const png = `data:image/png;base64,${canvas.toBuffer('image/png').toString('base64')}`;

// --- 4. upload rendered image (the render gate)
const up = await j(`/api/runs/${id}/image`, { method: 'POST', body: JSON.stringify({ image: png }) });
console.log('4. upload:', up.status, JSON.stringify(up.body).slice(0, 200));
if (!up.ok) process.exit(1);

// --- 5. wait for analysis
let run = null;
const t1 = Date.now();
while (Date.now() - t1 < 360000) {
  const r = await j(`/api/runs/${id}`);
  run = r.body;
  if (run.status === 'done' || run.status === 'error') break;
  await new Promise((res) => setTimeout(res, 4000));
}
console.log('5. final status:', run.status, run.error || '');
for (const st of run.stages || []) {
  console.log(`   - ${st.key}: ${st.status} (${st.ms}ms) ${String(st.output || '').slice(0, 90).replace(/\n/g, ' ')}`);
}
if (run.status !== 'done') process.exit(1);

// --- 6. verify the DB image is the real drawing (not blank)
const img = await fetch(`${BASE}/api/runs/${id}/image?api_key=${API_KEY}`);
const buf = Buffer.from(await img.arrayBuffer());
console.log('6. stored image:', img.status, img.headers.get('content-type'), buf.length, 'bytes');
process.exit(0);

File: e2e_test.py
# E2E test: agent draw -> render plan with PIL (same semantics as the
# browser replay) -> upload -> analysis -> verify stored image is real.
# Usage: DRAW_API_KEY=... python3 e2e_test.py "a small sailboat at sunset"
import json
import os
import sys
import time
import urllib.request
import urllib.error

BASE = os.environ.get("BASE", "http://127.0.0.1:8910")
API_KEY = os.environ["DRAW_API_KEY"]
subject = sys.argv[1] if len(sys.argv) > 1 else "a small sailboat on calm water at sunset"


def call(path, method="GET", data=None):
    req = urllib.request.Request(
        BASE + path,
        method=method,
        data=json.dumps(data).encode() if data is not None else None,
        headers={"content-type": "application/json", "x-api-key": API_KEY},
    )
    try:
        with urllib.request.urlopen(req, timeout=120) as r:
            return r.status, json.loads(r.read().decode() or "{}")
    except urllib.error.HTTPError as e:
        try:
            return e.code, json.loads(e.read().decode() or "{}")
        except Exception:
            return e.code, {}


st, create = call("/api/draw", "POST", {"prompt": subject, "stages": ["describe", "critique"]})
print("1. create:", st, create)
assert st == 200, create
run_id = create["run_id"]

plan = None
t0 = time.time()
while time.time() - t0 < 360:
    st, r = call(f"/api/runs/{run_id}")
    if r.get("status") == "error":
        print("PLAN ERROR:", r.get("error"))
        sys.exit(1)
    if r.get("status") == "rendering":
        ds = next((s for s in r.get("stages", []) if s.get("key") == "draw"), None)
        if ds and ds.get("plan"):
            plan = ds["plan"]
            break
    time.sleep(4)
assert plan, "no plan after 6 min"
print(f"2. plan ready in {time.time()-t0:.0f}s: {plan['count']} strokes, subject: {plan.get('subject')}")

# --- render the plan (same geometry the browser replays)
from PIL import Image, ImageDraw, ImageFont

W, H = 960, 600
img = Image.new("RGB", (W, H), "white")
dr = ImageDraw.Draw(img)
try:
    font_cache = {}
    def font(size):
        if size not in font_cache:
            try:
                font_cache[size] = ImageFont.truetype("DejaVuSans.ttf", size)
            except Exception:
                font_cache[size] = ImageFont.load_default()
        return font_cache[size]
except Exception:
    font = lambda s: ImageFont.load_default()

for s in plan["strokes"]:
    c, w = s.get("color", "#1a1a1a"), s.get("width", 3) or 3
    if s["type"] == "line":
        pts = [(p[0], p[1]) for p in s["points"]]
        if len(pts) >= 2:
            dr.line(pts, fill=c, width=w, joint="curve")
    elif s["type"] == "circle":
        x0, y0 = s["cx"] - s["r"], s["cy"] - s["r"]
        x1, y1 = s["cx"] + s["r"], s["cy"] + s["r"]
        if s.get("fill"):
            dr.ellipse([x0, y0, x1, y1], fill=c)
        else:
            dr.ellipse([x0, y0, x1, y1], outline=c, width=w)
    elif s["type"] == "rect":
        box = [s["x"], s["y"], s["x"] + s["w"], s["y"] + s["h"]]
        if s.get("fill"):
            dr.rectangle(box, fill=c)
        else:
            dr.rectangle(box, outline=c, width=w)
    elif s["type"] == "text":
        dr.text((s["x"], s["y"]), s.get("text", ""), fill=c, font=font(s.get("size", 24) or 24))
img.save("/tmp/e2e-render.png")

import base64
raw = open("/tmp/e2e-render.png", "rb").read()
png = "data:image/png;base64," + base64.b64encode(raw).decode()
print(f"3. rendered {len(raw)} bytes")

st, up = call(f"/api/runs/{run_id}/image", "POST", {"image": png})
print("4. upload:", st, json.dumps(up)[:200])
assert st == 200, up

run = None
t1 = time.time()
while time.time() - t1 < 360:
    st, run = call(f"/api/runs/{run_id}")
    if run.get("status") in ("done", "error"):
        break
    time.sleep(5)
print("5. final status:", run.get("status"), run.get("error") or "")
for s in run.get("stages", []):
    print(f"   - {s['key']}: {s['status']} ({s.get('ms')}ms) {str(s.get('output') or '')[:90].replace(chr(10), ' ')}")
assert run.get("status") == "done", run

imgreq = urllib.request.Request(f"{BASE}/api/runs/{run_id}/image?api_key={API_KEY}")
with urllib.request.urlopen(imgreq) as r:
    body = r.read()
    print("6. stored image:", r.status, r.headers.get("content-type"), len(body), "bytes")
    assert len(body) > 5000, "stored image too small — likely blank"
print("E2E PASS")
